src/monmove.cpp

```diff
@@ -1381,6 +1527,73 @@ bool monster::attack_at( const tripoint &p )
     return false;
 }
 
+bool monster::pickup_at( const tripoint &p, item_location &target )
+{
+    //If called through normal pathfinding, then clear looting effect
+    if( p == goal ) {
+        remove_effect( effect_looting );
+
+    }
+
+    //Item has moved...
+    if( p != target.position() ) {
+        return false;
+    }
+
+    inv.push_back( *target.get_item() );
+    item *stored_item = &inv.back();
+
+    //Get weight of the stack or item
+    units::mass weight = target->weight();
+    units::mass capacity = this->weight_capacity();
+
+    int amount_taken = 1;
+
+    int charges = target->charges;
+
+    if( charges >= 1 ) {
+        //Adjust volume and weight for units in a stack
+        units::mass weigh_each = weight / charges;
+        amount_taken = charges + ( weigh_each / capacity );
```

Comment:
Please make sure `capacity` isn't equals to zero, otherwise we'll get a division by zero here.

Reply:
Are you sure about this calculation? `/ capacity` means the stronger the monster is (= higher weight capacity), so **less** charges it will pick up.

Also: why are you adding this value to the existing charges of the item. The result can only be larger than the number of existing charges, so the code will try to pick up more than there is.

```C++
amount_taken = std::min( target->charges, capacity / weight_each );
```
The `std::min` is so it selects the smaller value of the amount of existing charges or the amount of charges that can be picked.